Hmm, one edge case that isn't currently covered in the tests is call-based `TypedDict`s. I'm not sure if this behaviour is correct:

```pycon
>>> from typing import TypedDict
>>> class Foo(TypedDict): ...
...
>>> Foo2 = TypedDict("Foo2", {})
>>> Foo.__orig_bases__
(<function TypedDict at 0x000002796B986410>,)
>>> Foo2.__orig_bases__
()
```

Here's the behaviour with call-based `NamedTuple`s, which is the closest analogue:

```pycon
>>> from typing import NamedTuple
>>> class Bar(NamedTuple): ...
...
>>> Bar2 = NamedTuple("Bar2", {})
>>> Bar.__orig_bases__
(<function NamedTuple at 0x000002796B986200>,)
>>> Bar2.__orig_bases__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Bar2' has no attribute '__orig_bases__'
```

--- a/Lib/test/test_typing.py
+++ b/Lib/test/test_typing.py
@@ -6695,6 +6695,46 @@ class NamedTupleTests(BaseTestCase):
                 self.assertEqual(jane2, jane)
                 self.assertIsInstance(jane2, cls)
 
+    def test_orig_bases(self):
+        T = TypeVar('T')
+
+        class Parent(NamedTuple):
+            pass
+
+        class Child(Parent):
+            pass
+
+        class OtherChild(Parent):
+            pass
+
+        class MixedChild(Child, OtherChild, Parent):
+            pass
+
+        class GenericParent(NamedTuple, Generic[T]):
+            pass
+
+        class GenericChild(GenericParent[int]):
+            pass
+
+        class OtherGenericChild(GenericParent[str]):
+            pass
+
+        class MixedGenericChild(GenericChild, OtherGenericChild, GenericParent[float]):
+            pass
+
+        class MultipleGenericBases(GenericParent[int], GenericParent[float]):
+            pass
+
+        self.assertEqual(Parent.__orig_bases__, (NamedTuple,))
+        self.assertEqual(Child.__orig_bases__, (Parent,))
+        self.assertEqual(OtherChild.__orig_bases__, (Parent,))
+        self.assertEqual(MixedChild.__orig_bases__, (Child, OtherChild, Parent,))
+        self.assertEqual(GenericParent.__orig_bases__, (NamedTuple, Generic[T]))
+        self.assertEqual(GenericChild.__orig_bases__, (GenericParent[int],))
+        self.assertEqual(OtherGenericChild.__orig_bases__, (GenericParent[str],))
+        self.assertEqual(MixedGenericChild.__orig_bases__, (GenericChild, OtherGenericChild, GenericParent[float]))
+        self.assertEqual(MultipleGenericBases.__orig_bases__, (GenericParent[int], GenericParent[float]))
+
 
 class TypedDictTests(BaseTestCase):
     def test_basics_functional_syntax(self):

--- a/Lib/test/test_typing.py
+++ b/Lib/test/test_typing.py
@@ -6698,42 +6698,18 @@ class NamedTupleTests(BaseTestCase):
     def test_orig_bases(self):
         T = TypeVar('T')
 
-        class Parent(NamedTuple):
+        class SimpleNamedTuple(NamedTuple):
             pass
 
-        class Child(Parent):
+        class GenericNamedTuple(NamedTuple, Generic[T]):
             pass
 
-        class OtherChild(Parent):
-            pass
+        self.assertEqual(SimpleNamedTuple.__orig_bases__, (NamedTuple,))
+        self.assertEqual(GenericNamedTuple.__orig_bases__, (NamedTuple, Generic[T]))
 
-        class MixedChild(Child, OtherChild, Parent):
-            pass
+        CallNamedTuple = NamedTuple('CallNamedTuple', [])
 
-        class GenericParent(NamedTuple, Generic[T]):
-            pass
-
-        class GenericChild(GenericParent[int]):
-            pass
-
-        class OtherGenericChild(GenericParent[str]):
-            pass
-
-        class MixedGenericChild(GenericChild, OtherGenericChild, GenericParent[float]):
-            pass
-
-        class MultipleGenericBases(GenericParent[int], GenericParent[float]):
-            pass
-
-        self.assertEqual(Parent.__orig_bases__, (NamedTuple,))
-        self.assertEqual(Child.__orig_bases__, (Parent,))
-        self.assertEqual(OtherChild.__orig_bases__, (Parent,))
-        self.assertEqual(MixedChild.__orig_bases__, (Child, OtherChild, Parent,))
-        self.assertEqual(GenericParent.__orig_bases__, (NamedTuple, Generic[T]))
-        self.assertEqual(GenericChild.__orig_bases__, (GenericParent[int],))
-        self.assertEqual(OtherGenericChild.__orig_bases__, (GenericParent[str],))
-        self.assertEqual(MixedGenericChild.__orig_bases__, (GenericChild, OtherGenericChild, GenericParent[float]))
-        self.assertEqual(MultipleGenericBases.__orig_bases__, (GenericParent[int], GenericParent[float]))
+        self.assertEqual(CallNamedTuple.__orig_bases__, (NamedTuple,))
 
 
 class TypedDictTests(BaseTestCase):
@@ -7196,6 +7172,8 @@ class TypedDictTests(BaseTestCase):
         class MultipleGenericBases(GenericParent[int], GenericParent[float]):
             pass
 
+        CallTypedDict = TypedDict('CallTypedDict', {})
+
         self.assertEqual(Parent.__orig_bases__, (TypedDict,))
         self.assertEqual(Child.__orig_bases__, (Parent,))
         self.assertEqual(OtherChild.__orig_bases__, (Parent,))
@@ -7205,6 +7183,7 @@ class TypedDictTests(BaseTestCase):
         self.assertEqual(OtherGenericChild.__orig_bases__, (GenericParent[str],))
         self.assertEqual(MixedGenericChild.__orig_bases__, (GenericChild, OtherGenericChild, GenericParent[float]))
         self.assertEqual(MultipleGenericBases.__orig_bases__, (GenericParent[int], GenericParent[float]))
+        self.assertEqual(CallTypedDict.__orig_bases__, (TypedDict,))
 
 
 class RequiredTests(BaseTestCase):

--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -2962,7 +2962,9 @@ def NamedTuple(typename, fields=None, /, **kwargs):
     elif kwargs:
         raise TypeError("Either list of fields or keywords"
                         " can be provided to NamedTuple, not both")
-    return _make_nmtuple(typename, fields, module=_caller())
+    nt = _make_nmtuple(typename, fields, module=_caller())
+    nt.__orig_bases__ = (NamedTuple,)
+    return nt
 
 _NamedTuple = type.__new__(NamedTupleMeta, 'NamedTuple', (), {})
 
@@ -3107,7 +3109,9 @@ def TypedDict(typename, fields=None, /, *, total=True, **kwargs):
         # Setting correct module is necessary to make typed dict classes pickleable.
         ns['__module__'] = module
 
-    return _TypedDictMeta(typename, (), ns, total=total)
+    td = _TypedDictMeta(typename, (), ns, total=total)
+    td.__orig_bases__ = (TypedDict,)
+    return td
 
 _TypedDict = type.__new__(_TypedDictMeta, 'TypedDict', (), {})
 TypedDict.__mro_entries__ = lambda bases: (_TypedDict,)

